ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc

```diff
@@ -712,21 +716,80 @@ TFLiteInterpreter::cacheInOutTensorPtr ()
 /**
  * @brief	TFLiteCore constructor
  */
-TFLiteCore::TFLiteCore ()
+TFLiteCore::TFLiteCore (const GstTensorFilterProperties *prop)
 {
   num_threads = -1;
   accelerator = ACCL_NONE;
   delegate = TFLITE_DELEGATE_NONE;
   interpreter_sub = nullptr;
-  interpreter = new TFLiteInterpreter ();
+  shared_tensor_filter_key = NULL;
+
+  if (prop->shared_tensor_filter_key) {
+    shared_tensor_filter_key =
+        g_strdup (prop->shared_tensor_filter_key);
+    checkSharedInterpreter (prop);
+  }
+  else
+    interpreter = new TFLiteInterpreter ();
+}
+
+/**
+ * @brief callback method to delete interpreter for shared model
+ * @param instance TFLiteCore* self instance pointer
+ */
+void free_interpreter (void * instance) {
+  TFLiteCore * core = reinterpret_cast <TFLiteCore *> (instance);
+  delete core->interpreter;
 }
 
 /**
  * @brief	TFLiteCore destructor
  */
 TFLiteCore::~TFLiteCore ()
 {
-  delete interpreter;
+  if (shared_tensor_filter_key) {
+    if (!nnstreamer_filter_shared_table_remove (shared_tensor_filter_key, this, free_interpreter)) {
+      nns_loge ("failed to remove shared model");
+    }
+    g_free (shared_tensor_filter_key);
+  }
+  else {
+    delete interpreter;
+  }
+}
+
+/**
+ * @brief	check the shared interpreter
+ * The shared model representation (interpreter) is allocated or shared.
+ * If `shared_tensor_filter_key` is already existed, it will share the TFLiteInterpreter.
+ * in the opposite case, the new TFLiteInterpreter will allocated and registered at the shared table.
+ */
+void
+TFLiteCore::checkSharedInterpreter (const GstTensorFilterProperties * prop)
+{
+  TFLiteInterpreter * ref = (TFLiteInterpreter *) nnstreamer_filter_get_shared_model_representation (shared_tensor_filter_key);
+  if (ref) {
+    if (g_strcmp0 (prop->model_files[0],
+            ref->getModelPath ()) != 0) {
+      ml_loge ("The model paths are different between the tensor filters!");
+      return;
+    }
+    if (!nnstreamer_filter_shared_table_insert (shared_tensor_filter_key, NULL, this)) {
+      ml_loge ("Failed to share the model representation!");
+      return;
+    }
+    interpreter = ref;
+  }
+  else {
+    interpreter = new TFLiteInterpreter ();
+    if (!nnstreamer_filter_shared_table_insert (shared_tensor_filter_key, interpreter, this)) {
+      ml_loge ("Failed to share the model representation!");
+      delete interpreter;
+      return;
+    }
+  }
```

Comment:
You still have thread-synchronization issue and calling get and then insert is not a good design.
I'd suggest:

```C
///this function
{
   ...
  interpreter = get (key);
  if (!interpreter) { // not available. create interpreter
    interpreter = insert_and_get (key, new-interpreter());
  }
  ...
}

...

///remove function
{
  put (key); // if internal counter or linked-list hits 0, it'll be removed by tensor-filter infra.
}
```

Reply:
That helps me understand how to synchronize. With that API we might need static mutex to lock around `get ()` and `insert_and_get ()`, right? Just in case remove function runs between the two